test/dom-parsing/XMLSerializer.tests.ts

```diff
@@ -0,0 +1,346 @@
+	});
+
+	it('always generates a new prefix for namespaced attributes', () => {
+		// TODO: this sounds like it is not intended behavior, but does follow the algorithm
+		// (see XML serialization of the attributes or an Element, 3.5.3.1.)
+		const el = document.createElement('test');
+		el.setAttributeNS('http://www.example.com/ns', 'prf:test', 'value');
+		chai.assert.equal(
+			serializer.serializeToString(el),
+			'<test xmlns:ns1="http://www.example.com/ns" ns1:test="value"/>'
+		);
+	});
+
+	it('always uses prefix xml for the xml namespace, even if declared as default', () => {
+		const el = document.createElementNS('http://www.example.com/ns', 'prf:test');
+		el.setAttributeNS(XMLNS_NAMESPACE, 'xmlns', XML_NAMESPACE);
+		el.appendChild(document.createElementNS(XML_NAMESPACE, 'test'));
+		chai.assert.equal(
+			serializer.serializeToString(el),
+			'<prf:test xmlns:prf="http://www.example.com/ns"><xml:test/></prf:test>'
+		);
+	});
+
+	it('ignores any namespace declarations for the xml namespace', () => {
+		const el = document.createElementNS(XML_NAMESPACE, 'test');
+		el.setAttributeNS(XMLNS_NAMESPACE, 'xmlns', XML_NAMESPACE);
+		el.setAttributeNS(XMLNS_NAMESPACE, 'xmlns:xml', XML_NAMESPACE);
+		el.setAttributeNS(XMLNS_NAMESPACE, 'xmlns:prf', XML_NAMESPACE);
+		chai.assert.equal(serializer.serializeToString(el), '<xml:test/>');
+	});
+
+	it('uses a HTML parser compatible serialization for empty HTML elements', () => {
+		const el = document.createElementNS(HTML_NAMESPACE, 'body');
+		el.appendChild(document.createElementNS(HTML_NAMESPACE, 'br'));
+		el.appendChild(document.createElementNS(HTML_NAMESPACE, 'i'));
+		el.appendChild(document.createElement('not-html'));
+		chai.assert.equal(
+			serializer.serializeToString(el),
+			'<body xmlns="http://www.w3.org/1999/xhtml"><br /><i></i><not-html xmlns=""/></body>'
+		);
+	});
+
+	it('can serialize a ProcessingInstruction', () => {
+		chai.assert.equal(
+			serializer.serializeToString(document.createProcessingInstruction('target', 'data')),
+			'<?target data?>'
+		);
+	});
+
+	it('can serialize a Text node', () => {
+		chai.assert.equal(serializer.serializeToString(document.createTextNode('test')), 'test');
+	});
+
+	it('can serialize an XMLDocument', () => {
+		chai.assert.equal(
+			serializer.serializeToString(document.implementation.createDocument('http://www.example.com/ns', 'test')),
+			'<test xmlns="http://www.example.com/ns"/>'
+		);
+	});
+
+	it("throws if given something that isn't a node", () => {
+		chai.assert.throws(() => (serializer as any).serializeToString({ nodeType: 1 }), TypeError);
+	});
+});
+
+describe('serializeToWellFormedString', () => {
+	let document: slimdom.Document;
+	beforeEach(() => {
+		document = new slimdom.Document();
+	});
+
+	it("throws if given something that isn't a node", () => {
+		chai.assert.throws(() => (slimdom as any).serializeToWellFormedString({ nodeType: 1 }), TypeError);
```

Comment:
This is a convenient way to disable type checking for the function call, otherwise TypeScript wouldn't let me pass the object as a Node.